### DISCUSSION TITLE
STM32F765: Different USB serial number in DFU and APP modes

### DISCUSSION THREAD

**Discussion by IlorDash** [micropython/micropython#15948] [STM32 / Pyboard]
STM32F765: Different USB serial number in DFU and APP modes

I have several STM32F765, probably bought them in 2020, that I used on my boards. They have a different USB serial number in DFU and APP modes. Serial numbers consistently different in fourth byte: in DFU mode it's greater by 0xA for each MCU. 

For example serial number in DFU mode - 356F345**D**3438:

```
usb 1-2: New USB device found, idVendor=0483, idProduct=df11, bcdDevice=22.00
usb 1-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
usb 1-2: Product: STM32  BOOTLOADER
usb 1-2: Manufacturer: STMicroelectronics
usb 1-2: SerialNumber: 356F345D3438
```
Serial number in APP mode - 356F345**3**3438:

```
usb 1-2: New USB device found, idVendor=f055, idProduct=9800, bcdDevice= 2.00
usb 1-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
usb 1-2: Product: My-Board-Name
usb 1-2: Manufacturer: MicroPython
usb 1-2: SerialNumber: 356F34533438
```

I also tried to read serial number in REPL:

```
>>> import machine
>>> UID_BASE = 0x1ff0f420
>>> uid0 = machine.mem32[UID_BASE]
>>> uid1 = machine.mem32[UID_BASE+4]
>>> uid2 = machine.mem32[UID_BASE+8]
>>> unique_id = (uid2 << 64) | (uid1 << 32) | uid0
>>> uid_bytes = [
... (uid0 >> 0) & 0xFF, (uid0 >> 8) & 0xFF, (uid0 >> 16) & 0xFF, (uid0 >> 24) & 0xFF,
... (uid1 >> 0) & 0xFF, (uid1 >> 8) & 0xFF, (uid1 >> 16) & 0xFF, (uid1 >> 24) & 0xFF,
... (uid2 >> 0) & 0xFF, (uid2 >> 8) & 0xFF, (uid2 >> 16) & 0xFF, (uid2 >> 24) & 0xFF,
... ]
>>> print(f"{uid_bytes[11]:X}{(uid_bytes[10]+uid_bytes[2]):X}{uid_bytes[9]:X}{(uid_bytes[8]+uid_bytes[0]):X}{uid_bytes[7]:X}{uid_bytes[6]:X}")
356F34533438
```

It's equal to the serial number in APP mode.

I found serial number algorithm description here - https://github.com/micropython/micropython/blob/17d82344581ad3a76033fae54c5d3304e17f185f/ports/stm32/usbd_desc.c#L132 But link to ST documents doesn't work.

So I'm wondering, if these bits can never be altered by the user, could the STM32 bootloader code have changed between MCU revisions, so it calculated the serial number differently?

**Comment by bidrohini** [comment]
Have you tried to hard-code the serial number in your application?

**Reply by IlorDash** [reply]
Yes, I'm able to hard-code DFU serial number in Micropython stm32 port sources, but it isn't what I am asking about.

My question is about default algorithm to calculate serial number in bootloader and MicroPython firmwares. Since MicroPython should use the same approach to compose serial number, as it's done in STM32 bootloader firmware, I expected serial numbers will be same in both modes.
